File: Traditional-Important-Algorithms/exercises_py/01_gc_and_memory/gc07_concurrent_gc.py
# I AM NOT DONE

"""
Exercise: Concurrent Garbage Collection

Concurrent GC runs the collector concurrently with the mutator (application).
This requires synchronization and write barriers to maintain correctness.

Your task: Implement a simplified concurrent GC with basic write barrier.
"""

import threading


class Object:
    """Object with references and mark bit"""

    def __init__(self, references=None):
        self.references = references if references is not None else []
        self.marked = False


class ConcurrentGC:
    """Concurrent garbage collector with write barrier"""

    def __init__(self):
        self.objects = {}  # obj_id -> Object
        self.next_id = 0
        self.roots = set()
        self.write_barrier_log = []  # Log of (from_id, to_id) writes
        self.lock = threading.Lock()

    def allocate(self, references=None):
        """Allocate a new object"""
        with self.lock:
            obj_id = self.next_id
            self.next_id += 1

            obj = Object(references if references else [])
            self.objects[obj_id] = obj
            return obj_id

    def add_root(self, obj_id):
        """Add an object as a root"""
        with self.lock:
            self.roots.add(obj_id)

    def write_barrier(self, from_id, to_id):
        """Record a pointer write"""
        # TODO: Implement write barrier
        # Log pointer writes during concurrent marking
        # This ensures we don't miss references created during GC
        pass

    def update_reference(self, from_id, index, to_id):
        """Update a reference with write barrier"""
        # TODO: Update a reference with write barrier
        # 1. Call write barrier
        # 2. Update the actual reference
        pass

    def _mark(self, obj_id, marked):
        """Mark an object and its references"""
        if obj_id in marked:
            return

        marked.add(obj_id)

        with self.lock:
            if obj_id in self.objects:
                obj = self.objects[obj_id]
                refs = obj.references.copy()

        for ref_id in refs:
            self._mark(ref_id, marked)

    def concurrent_collect(self):
        """Perform concurrent collection"""
        # TODO: Implement concurrent collection
        # 1. Mark from roots
        # 2. Process write barrier log
        # 3. Sweep unmarked objects
        pass

    def object_count(self):
        """Return number of live objects"""
        with self.lock:
            return len(self.objects)


import unittest


class TestConcurrentGC(unittest.TestCase):
    def test_concurrent_allocation(self):
        gc = ConcurrentGC()

        obj1 = gc.allocate()
        obj2 = gc.allocate([obj1])

        gc.add_root(obj2)

        gc.concurrent_collect()

        self.assertEqual(gc.object_count(), 2)

    def test_write_barrier(self):
        gc = ConcurrentGC()

        obj1 = gc.allocate()
        obj2 = gc.allocate()

        gc.add_root(obj1)

        gc.update_reference(obj1, 0, obj2)

        gc.concurrent_collect()

        self.assertEqual(gc.object_count(), 2)

    def test_concurrent_threads(self):
        gc = ConcurrentGC()

        obj1 = gc.allocate()
        gc.add_root(obj1)

        def collector_thread():
            gc.concurrent_collect()

        thread = threading.Thread(target=collector_thread)
        thread.start()

        # Allocate while GC runs
        obj2 = gc.allocate()

        thread.join()

        # At least obj1 should survive
        self.assertGreaterEqual(gc.object_count(), 1)

    def test_multiple_writes(self):
        gc = ConcurrentGC()

        obj1 = gc.allocate()
        obj2 = gc.allocate()
        obj3 = gc.allocate()

        gc.add_root(obj1)

        gc.update_reference(obj1, 0, obj2)
        gc.update_reference(obj1, 1, obj3)

        gc.concurrent_collect()

        self.assertEqual(gc.object_count(), 3)

    def test_no_roots(self):
        gc = ConcurrentGC()

        gc.allocate()
        gc.allocate()

        gc.concurrent_collect()

        self.assertEqual(gc.object_count(), 0)

    def test_transitive_marking(self):
        gc = ConcurrentGC()

        obj1 = gc.allocate()
        obj2 = gc.allocate([obj1])
        obj3 = gc.allocate([obj2])

        gc.add_root(obj3)

        gc.concurrent_collect()

        self.assertEqual(gc.object_count(), 3)

    def test_write_during_collection(self):
        gc = ConcurrentGC()

        obj1 = gc.allocate()
        obj2 = gc.allocate()

        gc.add_root(obj1)

        # Simulate write during collection
        gc.write_barrier(obj1, obj2)

        gc.concurrent_collect()

        # Both should survive due to write barrier
        self.assertEqual(gc.object_count(), 2)

    def test_multiple_roots(self):
        gc = ConcurrentGC()

        obj1 = gc.allocate()
        obj2 = gc.allocate()
        obj3 = gc.allocate()

        gc.add_root(obj1)
        gc.add_root(obj2)

        gc.concurrent_collect()

        self.assertEqual(gc.object_count(), 2)

    def test_complex_graph(self):
        gc = ConcurrentGC()

        obj1 = gc.allocate()
        obj2 = gc.allocate([obj1])
        obj3 = gc.allocate([obj1, obj2])
        obj4 = gc.allocate()

        gc.add_root(obj3)

        gc.concurrent_collect()

        self.assertEqual(gc.object_count(), 3)

    def test_sequential_collections(self):
        gc = ConcurrentGC()

        obj1 = gc.allocate()
        gc.add_root(obj1)

        gc.concurrent_collect()
        self.assertEqual(gc.object_count(), 1)

        gc.concurrent_collect()
        self.assertEqual(gc.object_count(), 1)

    def test_barrier_log_processing(self):
        gc = ConcurrentGC()

        obj1 = gc.allocate()
        obj2 = gc.allocate()

        gc.add_root(obj1)

        # Multiple writes
        gc.write_barrier(obj1, obj2)
        gc.write_barrier(obj1, obj2)

        gc.concurrent_collect()

        self.assertEqual(gc.object_count(), 2)


if __name__ == '__main__':
    unittest.main()

File: Traditional-Important-Algorithms/exercises/01_gc_and_memory/gc07_concurrent_gc.rs
// gc07_concurrent_gc.rs
//
// Concurrent GC runs the collector concurrently with the mutator (application).
// This requires synchronization and write barriers to maintain correctness.
//
// Your task: Implement a simplified concurrent GC with basic write barrier.

// I AM NOT DONE

use std::collections::{HashMap, HashSet};
use std::sync::{Arc, Mutex};

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub struct ObjectId(usize);

#[derive(Debug, Clone)]
struct Object {
    references: Vec<ObjectId>,
    marked: bool,
}

#[derive(Clone)]
pub struct ConcurrentGC {
    objects: Arc<Mutex<HashMap<ObjectId, Object>>>,
    next_id: Arc<Mutex<usize>>,
    roots: Arc<Mutex<HashSet<ObjectId>>>,
    write_barrier_log: Arc<Mutex<Vec<(ObjectId, ObjectId)>>>,
}

impl ConcurrentGC {
    pub fn new() -> Self {
        Self {
            objects: Arc::new(Mutex::new(HashMap::new())),
            next_id: Arc::new(Mutex::new(0)),
            roots: Arc::new(Mutex::new(HashSet::new())),
            write_barrier_log: Arc::new(Mutex::new(Vec::new())),
        }
    }

    pub fn allocate(&self, references: Vec<ObjectId>) -> ObjectId {
        let mut next_id = self.next_id.lock().unwrap();
        let id = ObjectId(*next_id);
        *next_id += 1;

        let obj = Object {
            references,
            marked: false,
        };

        let mut objects = self.objects.lock().unwrap();
        objects.insert(id, obj);
        id
    }

    pub fn add_root(&self, id: ObjectId) {
        let mut roots = self.roots.lock().unwrap();
        roots.insert(id);
    }

    pub fn write_barrier(&self, from: ObjectId, to: ObjectId) {
        // TODO: Implement write barrier
        // Log pointer writes during concurrent marking
        // This ensures we don't miss references created during GC
        todo!()
    }

    pub fn update_reference(&self, from: ObjectId, index: usize, to: ObjectId) {
        // TODO: Update a reference with write barrier
        // 1. Call write barrier
        // 2. Update the actual reference
        todo!()
    }

    fn mark(&self, id: ObjectId, marked: &mut HashSet<ObjectId>) {
        if marked.contains(&id) {
            return;
        }
        marked.insert(id);

        let objects = self.objects.lock().unwrap();
        if let Some(obj) = objects.get(&id) {
            let refs = obj.references.clone();
            drop(objects);

            for &ref_id in &refs {
                self.mark(ref_id, marked);
            }
        }
    }

    pub fn concurrent_collect(&self) {
        // TODO: Implement concurrent collection
        // 1. Mark from roots
        // 2. Process write barrier log
        // 3. Sweep unmarked objects
        todo!()
    }

    pub fn object_count(&self) -> usize {
        let objects = self.objects.lock().unwrap();
        objects.len()
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::thread;

    #[test]
    fn test_concurrent_allocation() {
        let gc = ConcurrentGC::new();

        let obj1 = gc.allocate(vec![]);
        let obj2 = gc.allocate(vec![obj1]);

        gc.add_root(obj2);

        gc.concurrent_collect();

        assert_eq!(gc.object_count(), 2);
    }

    #[test]
    fn test_write_barrier() {
        let gc = ConcurrentGC::new();

        let obj1 = gc.allocate(vec![]);
        let obj2 = gc.allocate(vec![]);

        gc.add_root(obj1);

        gc.update_reference(obj1, 0, obj2);

        gc.concurrent_collect();

        assert_eq!(gc.object_count(), 2);
    }

    #[test]
    fn test_concurrent_threads() {
        let gc = ConcurrentGC::new();

        let obj1 = gc.allocate(vec![]);
        gc.add_root(obj1);

        let gc_clone = gc.clone();
        let handle = thread::spawn(move || {
            gc_clone.concurrent_collect();
        });

        // Allocate while GC runs
        let _obj2 = gc.allocate(vec![]);

        handle.join().unwrap();

        assert!(gc.object_count() >= 1);
    }
}
